### PR TITLE
Rewrite in TypeScript

### LINKED ISSUE #336: Rewrite in TypeScript
I originally rewrote it in Flow to get a feel for it. These days I only use TypeScript, so I would prefer to make this be in TypeScript too.

PR welcome if you want the challenge.

### PR DESCRIPTION
**Pull request by cluk3**
Fixes #336

This is my first time with Typescript, I'm expecting a lot of requested changes :)

### CODE REVIEW

File: global.d.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+declare module '*';
```
Comment (sindresorhus) [inline global.d.ts]: Why this?

File: lib/index.ts [hunk 2]
```diff
@@ -22,63 +20,65 @@ import template from 'lodash.template';
 import plur from 'plur';
 import unusedFilename from 'unused-filename';
 
-type PageresStream = Readable & {filename: string};
+interface PageresStream extends Readable {
+	filename: string;
+}
```
Comment (sindresorhus) [inline lib/index.ts]: What's the benefit of using `interface` here? The type existing type annotation works in TS too.

File: lib/index.ts [hunk 3]
```diff
@@ -95,22 +95,28 @@ export default class Pageres<DestValue: string> extends EventEmitter {
...
+	src() : Src[];
+	src(url: string, sizes: string[], options?: Options) : this;
+	src(url?: string, sizes?: string[], options?: Options) : this | Src[] {
 		if (url === undefined) {
 			return this._src;
 		}
 
+		if (sizes === undefined) {
+			throw new TypeError('Sizes required');
+		}
```
Comment (sindresorhus) [inline lib/index.ts]: This would be better enforced by the type-system.

File: lib/index.ts [hunk 4]
```diff
@@ -120,15 +126,15 @@ export default class Pageres<DestValue: string> extends EventEmitter {
 	}
 
 	async run(): Promise<PageresStream[]> {
-		await Promise.all(this.src().map(src => { // eslint-disable-line array-callback-return
-			const options = {...this.options, ...src.options};
-			const sizes = arrayUniq(src.sizes.filter(/./.test, /^\d{2,4}x\d{2,4}$/i));
-			const keywords = arrayDiffer(src.sizes, sizes);
-
+		await Promise.all(this.src().map((src : Src) : Promise<void> | void => {
```
Comment (sindresorhus) [inline lib/index.ts]:
```suggestion
		await Promise.all(this.src().map((src: Src) : Promise<void> | void => {
```

File: lib/index.ts [hunk 5]
```diff
@@ -176,8 +182,8 @@ export default class Pageres<DestValue: string> extends EventEmitter {
 		}
 	}
 
-	save(streams: Array<PageresStream>) {
-		const files = [];
+	async save(streams: PageresStream[]) {
+		const files : any[] = [];
```
Comment (sindresorhus) [inline lib/index.ts]: Don't use `any`.

File: lib/index.ts [hunk 6]
```diff
@@ -206,7 +211,7 @@ export default class Pageres<DestValue: string> extends EventEmitter {
 				});
 
 				write.on('finish', resolve);
-				write.on('error', async err => {
+				write.on('error', async (err : any) => {
```
Comment (sindresorhus) [inline lib/index.ts]: Don't use `any`.

File: package.json [hunk 7]
```diff
@@ -26,9 +26,10 @@
 		"node": ">=8"
 	},
 	"scripts": {
-		"test": "xo && npm run prepublish && nyc ava",
-		"flow": "flow || true",
-		"prepublish": "babel lib --out-dir=dist"
+		"test": "npm run lint && npm run build && nyc ava",
+		"lint": "tslint --format stylish --project . && xo",
```
Comment (sindresorhus) [inline package.json]: You can drop `xo`. No point in having both Xo and `tslint`.

File: package.json [hunk 8]
```diff
@@ -84,18 +85,11 @@
 		"viewport-list": "^5.0.1"
 	},
 	"devDependencies": {
+		"@types/node": "^10.12.0",
```
Comment (sindresorhus) [inline package.json]: Don't forget to import types for all the node modules. See: https://github.com/sindresorhus/pageres/pull/342/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R88

File: tsconfig.json [hunk 9]
```diff
@@ -0,0 +1,33 @@
+{
+	"compilerOptions": {
+		"outDir": "dist",
+		"target": "es2016",
+		"lib": [
+			"es2016",
+			"es2017.sharedmemory",
+			"esnext.asynciterable",
+			"dom"
```
Comment (sindresorhus) [inline tsconfig.json]: I think you only need `es2016` out of these.

File: global.d.ts (hunk 1 above)
Comment (joshghent) [reply global.d.ts]: This was due to a number of modules not having type definitions. It's a bit unsafe though, ideally we should declare each package that doesn't have type definitions (or maybe look to contribute back to them)

File: lib/index.ts (hunk 5 above)
Comment (joshghent) [reply lib/index.ts]: Would it be worth enforcing this in the tsconfig?

Comment (sindresorhus) [reply lib/index.ts]: Yes, but hard to enable `noImplicitAny` if you don't control all the dependencies. But we could try, sure.

**Comment by sindresorhus** [comment]
Friendly ping, @cluk3 :)

**Pull request merged**